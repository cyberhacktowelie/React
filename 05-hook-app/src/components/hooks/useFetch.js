import { useState , useEffect, useRef } from 'react';

export const useFetch = ( url ) => {

    const [state, setState] = useState({data:null, loading:true, error:null});

    const isMounted = useRef(true);

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, []);


    useEffect(() => {
        setState({data:null, loading:true, error:null});

        fetch( url )
        .then(response => response.json())
        .then(data => {
            setTimeout(() => {
                if(isMounted.current){
                    setState({data, loading:false, error:null});
                }
            },4000);
        }
        )
        .catch(error => {
            setState({data:null, loading:false, error:error});
        }
        )
    }, [url]);

    return state;

}
