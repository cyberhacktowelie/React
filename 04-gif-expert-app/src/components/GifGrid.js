// import React, { useState, useEffect } from 'react';
import { PropTypes } from 'prop-types';

import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';
import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ category }) => {

    // const [ images, setImages ] = useState([]);

    const { data:images , loading } = useFetchGifs(category);

    // useEffect(() => {
    //     // getGifs();
    //     // getGifs(category).then(img => setImages(img));  
    //     getGifs(category).then(setImages);
    // }, [ category ]);


    // const apiKey = 'QKXFCiGkEFDCSb1fyIPKGNNZ2wIx1E5g';
    // const getGifs = async () => {
    //     const url = `https://api.giphy.com/v1/gifs/search?q=${ encodeURI(category) }&limit=10&api_key=${apiKey}`;
    //     const resp = await fetch(url);
    //     const { data } = await resp.json();

    //     const gifs = data.map(img => {
    //         return {
    //             id: img.id,
    //             title: img.title,
    //             url: img.images?.downsized_medium.url
    //         }});

    //     // console.log(gifs);
    //     setImages(gifs);
    // };

    // getGifs();

    
    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {/* { loading ? <p>Cargando...</p> : null } */}
            { loading && <p className="animate__animated animate__flash">Cargando...</p> }
            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem key={ img.id} {...img} />
                    ))
                }
            </div>
        </>
    );
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}