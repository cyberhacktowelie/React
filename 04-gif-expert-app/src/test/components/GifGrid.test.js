import '@testing-library/jest-dom';
import { GifGrid } from '../../components/GifGrid';
import { shallow } from 'enzyme';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente GifGrid   GifGrid.test.js', () => {

    const category = 'One Punch';

    test('Debe de mostrarse correctamente', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}/>);

        expect( wrapper ).toMatchSnapshot();
    })

    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        
        const gifs = [
            {
                id: 'ABC',
                url: 'https://localhost/cualquiercosa.jpg',
                title: 'Cualquier cosa'
            },
            {
                id: 'DEF',
                url: 'https://localhost/otracosa.jpg',
                title: 'Otra cosa'
            }   
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category}/>);

        // expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find('p').exists() ).toBe(false);

        expect( wrapper.find('GifGridItem').length ).toBe(gifs.length);


    })

});