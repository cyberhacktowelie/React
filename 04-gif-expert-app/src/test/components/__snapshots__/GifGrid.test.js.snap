// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente GifGrid   GifGrid.test.js Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
